distributed/scheduler.py

```diff
@@ -6216,16 +6216,25 @@ def adaptive_target(self, comm=None, target_duration=None):
             cpu = max(1, cpu)
 
         # Memory
-        limit_bytes = {addr: ws._memory_limit for addr, ws in self.workers.items()}
-        worker_bytes = [ws._nbytes for ws in self.workers.values()]
-        limit = sum(limit_bytes.values())
-        total = sum(worker_bytes)
-        if total > 0.6 * limit:
+        limit = sum(ws.memory_limit for ws in self.workers.values())
+        total = sum(ws.nbytes for ws in self.workers.values())
+        if total < 0.6 * limit or limit <= 0:
             memory = 2 * len(self.workers)
         else:
             memory = 0
 
         target = max(memory, cpu)
+
+        # runnable tasks
+        tasks = 0
+        for task in self.tasks.values():
```

Comment:
There may be millions of tasks.  In general we can't iterate through all tasks every time we want to ask for the target.  Instead I recommend that you iterate through `TaskGroup`s and look at the `.states` attribute.